### PR TITLE
added point collector

### PR DESCRIPTION
**Pull request by lgostling**

### CODE REVIEW

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:656]: Let's call this pointcollector. I know the rest of the GH queue don't have the collector suffix, but that would make things more consistent with ADO.

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:684]: Nit: string instead of String.

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:690]: Depending on the purpose of this cache, I would suggest calling it differently.

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:695]: Let's not give 95% of the usage to point collector, at least until all web hook processing is off webhook processor. Maybe start with 75-85%?

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:706]: Hmmmm, at some point, let's revisit how we are using AdlsClient to initiate the record writers. I think this is the case in ADO collectors as well, but considering that we don't do directly ingestion to ADLS anymore (within the collectors), we should make this simpler. Can you leave a comment here so that we don't forget?

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:737]: Alan kept fixing this for me, but using throw; instead of throw exception is the better approach. This way you don't generate another level of indirection in the stack trace :) See, Alan, I keep learning...

File: GitHub.Collectors/Cache/PointCollectorTableEntity.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Cache/PointCollectorTableEntity.cs:6]: Not sure how this cache works, could you please explain? I have a fear that this might not be doing what you are looking for.

Is this attempting to cache a particular request / record so that if we are already going to process it, we would not again? If this is the case, would not keeping track of the exact URL to call (instead of org. repo, record type) easier? Also, when do we remove stuff? After all, you want to re-collect some URLs over time in case something has changed. (This might come later in the PR)

File: GitHub.Collectors/PointCollectorInput.cs
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:14]: Can we make Context similar to what we have in ADO point collector, a dictionary of <string, JToken>?

File: GitHub.Collectors/PointCollectorInput.cs
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:12]: Do we have any instantiations of PointCollectorInput? If not, why have virtual stuff?

File: GitHub.Collectors/PointCollectorInput.cs
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:21]: getOrganizationLogin => GetOrganizationLogin

File: GitHub.Collectors/PointCollectorInput.cs
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:31]: getRepository => GetRepository.

File: GitHub.Collectors/Processor/PointProcessor.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:63]: Do we really need  a new output just to change the output name wrt record type? I think it is fine for this to stay as Point_<blah>, what do you think?

File: GitHub.Collectors/Processor/PointProcessor.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:78]: Are you planning to implement these before merging, or do you plan to bring those in a separate PR? Regardless, as I mentioned above, I don't think this cache is working as you hope for. 

If I recall our discussion correctly, we want to have a cache because we would like to ensure that we are not doing point collection on the same URL over and over, especially over a short amount of time. Here is the concrete problem:
* microsoft organization does updates to 1000 PRs.
* GH fires 1000 events where organization URL is: github.com/microsoft
* We add 1000 messages to the point collector queue with the same request.

I think the idea is that you don't even want to add these messages at the beginning. You can do this only by keeping a cache that tells you when you have collected a particular URL latest (this to be updated / inserted by the point collector) and then not redirecting the collection to the point collector if this URL was collected recently (let's say within the last 5 minutes). This check and skip needs to be baked into Webhook processor.

In other words, point collector would be a producer for this cache whereas the webhook processor would be a consumer.

If you had something else in mind, please let me know. I might be mis-remembering.

**Review by kivancmuslu — COMMENTED** [review]
Good skeleton, missing stuff, but I am sure you are aware of this.

I really like the idea to wrap point collector into the abstract collector and treat this as a collector with a single web request. I think that simplifies a lot already.

I think the major feedback is on how cache is created and utilized. Please let me know if you have any questions.

**Comment by kivancmuslu** [comment]
@AllainPL please take a look to this when you get a chance to. Thanks!

File: GitHub.Collectors/Model/DefaultCollector.cs
Comment (kivancmuslu) [reply GitHub.Collectors/Model/DefaultCollector.cs:100]: ADO collectors are definitely doing this some way, if not AzureHelpers.

File: GitHub.Collectors/Model/DefaultCollector.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:71]: Similar to other caches, I would initialize this cache at higher-level e.g., functions class and then pass to Webhook processor and then DefaultCollector. This way, you don't keep initializing it (even if that is optimized for that).

File: GitHub.Collectors/Model/DefaultCollector.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:68]: >string responseType = "Array" [](http://example.com/codeflow?start=121&length=29)

Any reason for using "Array" as default but then using "Object" above for both. If there is no "good" default, then, leave it empty so that every called must decide what to pass.

File: GitHub.Collectors/Model/DefaultCollector.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:64]: >"Object" [](http://example.com/codeflow?start=138&length=8)

Can you please name this parameter e.g.,
```
responseType: "Object"
```

Otherwise the reviewer needs to go to the method eveytime to remember what it is used for.

File: GitHub.Collectors/Cache/PointCollectorTableEntity.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Cache/PointCollectorTableEntity.cs:13]: >this.PartitionKey = urlHash; [](http://example.com/codeflow?start=12&length=28)

I am assuming we do this mainly because Url has some characters that is not permitted to be a partition key? I wonder if this will cost us read-ability and debug-ability.

Can you potentially experiemnt with Url escaping to see whether it makes the URL safe enough to use as partition key? I am just thinking about making the table user-friendlly to explore in the future. Only if this won't take much time.

File: GitHub.Collectors/Model/DataContract.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DataContract.cs:8]: >OrgIstance [](http://example.com/codeflow?start=69&length=10)

Can you cross-check with Sascha to see how painful it would be to fix this e.g., typo here (Instance instead of Istance)?

File: GitHub.Collectors/Processor/PointProcessor.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:48]: >input.ResponseType == "Object" [](http://example.com/codeflow?start=15&length=30)

Nit: prefer to use .Equals(..) method of string instead of ==

File: GitHub.Collectors/Processor/PointProcessor.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:64]: Can you double check (code and/or end-to-end exeuciton) for the semantics of CacheAsync when the record already exists in the cache? Does it do a forced update? If that is the case, I think we are good, but good to check.

Comment (lgostling) [reply GitHub.Collectors/Processor/PointProcessor.cs:64]: Yes, I have tested that when the table entry already exists this performs an update.

File: GitHub.Collectors/Model/DefaultCollector.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:32]: >private set; [](http://example.com/codeflow?start=70&length=12)

Nit: if these are only set in the constructor, you can remove "private set;" parts.

File: GitHub.Collectors/Model/DataContract.cs
Comment (kivancmuslu) [reply GitHub.Collectors/Model/DataContract.cs:8]: Discussed with Sascha, we are not doing this yet. Please add a comment that at some point this needs to be fixed.

File: GitHub.Collectors/Model/CollectorFactory.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Model/CollectorFactory.cs:25]: >pointCache [](http://example.com/codeflow?start=73&length=10)

Maybe call this "pointcollectorcache"?

**Comment by kivancmuslu** [comment]
>             HttpResponseMessage response = await this.HttpClient.GetAsync(url, this.Authentication, DataContract.CommitInstanceApiName, allowlistedResponses).ConfigureAwait(false);

Are we planning to do this later on? Totally fine to leave it to a separate PR, just asking. Also, while doing this, can you please also clean-up passing / using HttpClient in Webhook processors and Default / Commit collectors? They should no longer need it (I think).

---
Refers to: GitHub.Collectors/Model/PushCollector.cs:109 in 880409e. [](commit_id = 880409ed13b3fedd0fdad4826396d83bf1da4f84, deletion_comment = False)

File: GitHub.Collectors/Processor/PointProcessor.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:16]: >PointProcessor [](http://example.com/codeflow?start=17&length=14)

Can we call this PointCollector instead of processor?

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:662]: Can you please also add a "dri" variant for this? As in ADO collectors.

**Comment by kivancmuslu** [comment]
This looks good on my end, please ensure that you get either Alan's or Sascha's (or both) feedback approval as well. In case I missed anything. Thanks!

---
In reply to: [751061162](https://github.com/microsoft/CEDAR.GitHub.Collector/pull/96#issuecomment-751061162) [](http://example.com/codeflow?ancestors=751061162,751061162)

**Review by kivancmuslu — APPROVED** [review]
:shipit:

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (AllainPL) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:704]: Is there a reason for this arbitrary number `80.0`?

File: GitHub.Collectors/Collector/PointCollector.cs
Comment (AllainPL) [inline GitHub.Collectors/Collector/PointCollector.cs:46]: NIT:
```suggestion
            Dictionary<string, JToken> additionalMetadata = new Dictionary<string, JToken>
            {
                { "OrganizationId", input.Repository.OrganizationId },
                { "OrganizationLogin", input.Repository.OrganizationLogin }
            };
```

File: GitHub.Collectors/Collector/PointCollector.cs
Comment (AllainPL) [inline GitHub.Collectors/Collector/PointCollector.cs:86]: Intentionally left? If yes ass some explanation :)

File: GitHub.Collectors/Model/CollectorFactory.cs
Comment (AllainPL) [inline GitHub.Collectors/Model/CollectorFactory.cs:28]: 🤟 I just passionately love this construction

File: GitHub.Collectors/Model/DefaultCollector.cs
Comment (AllainPL) [inline GitHub.Collectors/Model/DefaultCollector.cs:40]: string interpolation not needed

File: GitHub.Collectors/Collector/PointCollector.cs
Comment (AllainPL) [inline GitHub.Collectors/Collector/PointCollector.cs:55]: nit, spaces :)

**Review by AllainPL — COMMENTED** [review]
I would also review the `using` statements just at a quick glance it looks like some might be not used.

Btw we will need to sit down one day and create some automatic checks and rules + configure IDEs for everyone, as all the spaces using directives and even the Dictionary initialization should be autofixable/autodiscoverable

File: GitHub.Collectors/Model/PushCollector.cs
Comment (kivancmuslu) [inline GitHub.Collectors/Model/PushCollector.cs:29]: Naming convention: FunctionContext => functionContext

Comment (kivancmuslu) [reply GitHub.Collectors/Model/PushCollector.cs:29]: Make readonly as well?

File: GitHub.Collectors.Functions/GitHubFunctions.cs
Comment (AllainPL) [reply GitHub.Collectors.Functions/GitHubFunctions.cs:704]: @lgostling I'm still interested in this :P

**Pull request merged**
